### PR TITLE
Commands for managing Server and Database Roles

### PR DESCRIPTION
**Pull request by dbaduck**
## Type of Change

 - [ ] Bug fix (non-breaking change, fixes #<enter issue number>)
 - [X] New feature (non-breaking change, adds functionality)
 - [ ] Breaking change (effects multiple commands or functionality)
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 

### Purpose
 
Add gap functions for Roles

### Commands to test

New-DbaDbUser -SqlInstance localhost -Database mydb -Login user1
Add-DbaDbRoleMember -SqlInstance localhost -Database mydb -Role db_owner -DatabaseUser user1

Get-DbaDbRole -SqlInstance localhost -Database msdb

Remove-DbaDbRoleMember -SqlInstance localhost -Database mydb -Role db_owner -DatabaseUser user1

### CODE REVIEW

**Comment by potatoqualitee** [comment]
Thanks, duck! Will take a look when I have more time. 

In the meantime, if you open this repo in VS Code, it _should_ properly format your code so there are no failures there. If you do not want to use VSCode, please do `Get-ChildItem yourcommands | Invoke-DbatoolsFormatter` so that it converts to OTSB format. I'm unsure if it changes the tabs to spaces, though. If not, please replace all tabs with 4 spaces 💯 

Also @joshcorr wrote a function to autogen the unit tests for commands. Josh, can you walk Ben through it?

**Comment by sqllensman** [comment]
Hi @dbaduck 

I needed these commands for another purpose and have taken your code and updated it to include
Piping Support
Support for Should Process 
Unit Tests

My branch of your code is currently at https://github.com/sqllensman/dbatools/tree/DBADuck

**Comment by wsmelton** [comment]
@sqllensman you have contributor rights to this repo so you should be able to push your changes to this PR directly.

**Comment by dbaduck** [comment]
Thanks for letting me know and thanks for the mods.

Ben Miller

**Comment by potatoqualitee** [comment]
(all new commands will have to be published to the prerelease due to all of the changes we've made in psd1 and psm1)

**Comment by potatoqualitee** [comment]
i just realized what these commands are for and all ican say is helll yeahhh, i needed that just htis week!

**Comment by wsmelton** [comment]
Fixing formatting and removing some comments that are not "desired" 😁 

Still need to get test created but can do that in another PR if desired.

**Comment by potatoqualitee** [comment]
i cant find patrick's branch :( the tests have to be done or it'll fail the "must have one test" test

**Comment by wsmelton** [comment]
Looks like it was deleted.

**Comment by sqllensman** [comment]
I reset my fork to be based on PreRelease branch

Changes are in
https://github.com/sqllensman/dbatools/tree/DbaDuck

**Comment by sqllensman** [comment]
Pester tests don't currently work due to issue with .Query call in Pre-release branch

**Comment by potatoqualitee** [comment]
I will fix and ask @StartAutomating to give it a proper fix later, brb

**Comment by potatoqualitee** [comment]
adding `Update-TypeData $script:PSModuleRoot\xml\dbatools.Types.ps1xml` to psm1

**Comment by potatoqualitee** [comment]
k done, update and you should be set

**Comment by potatoqualitee** [comment]
please feel free to update duck's PR, btw, he said yes in dev or here and i'd love to get this merged 🙇 if not, please ping ben.

**Comment by sqllensman** [comment]
How do I do that?

**Comment by potatoqualitee** [comment]
try using branch `pr/5520` - i use the GUI and that is what the name appears as

image

**Comment by wsmelton** [comment]
If you use VS Code install the GitHub PR extension and it will handle checking out the PR branch for you.

**Comment by sqllensman** [comment]
I installed the  GitHub PR extension but I don't see anything that helps with this task. I tried assigning the Pull request, as that seemed to be one possible category but nothing changed

**Comment by wsmelton** [comment]
Right click and then select check out.

You can also find a button to check out in the description editor.

**Comment by wsmelton** [comment]
https://code.visualstudio.com/blogs/2018/09/10/introducing-github-pullrequests

**Comment by sqllensman** [comment]
Problem is that it doe not display any Pull requests for dbatools

I can see two old pull requests from dbareports but nothing fro dbatools

There is nothing to right-click on.

**Comment by wsmelton** [comment]
You have to open your local repo of dbatools.

**Comment by sqllensman** [comment]
This was not problem.
The problem is that the extension cannot handle having multiple repo's opened in a workspace.

After removing all but the dbatools repo then it started working

**Comment by potatoqualitee** [comment]
🍌 💃

File: functions/Get-DbaDbRole.ps1 [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                Add-Member -Force -InputObject $dbRole -MemberType NoteProperty -Name ComputerName -Value $server.ComputerName
+                Add-Member -Force -InputObject $dbRole -MemberType NoteProperty -Name InstanceName -Value $server.ServiceName
+                Add-Member -Force -InputObject $dbRole -MemberType NoteProperty -Name SqlInstance -Value $server.DomainInstanceName
+                Add-Member -Force -InputObject $dbRole -MemberType NoteProperty -Name Database -Value $db.Name
+
+                Select-DefaultView -InputObject $dbRole -Property "ComputerName", "InstanceName", "Database", "Name", "IsFixedRole"
+            }
+        }
+    }
+    end {
```
Comment (wsmelton) [inline functions/Get-DbaDbRole.ps1]: This needs to be removed.

File: functions/Remove-DbaDbRole.ps1 [hunk 2]
```diff
@@ -0,0 +1,156 @@
+#ValidationTags#CodeStyle, Messaging, FlowControl, Pipeline#
```
Comment (wsmelton) [inline functions/Remove-DbaDbRole.ps1]: Remove

File: functions/Remove-DbaDbRole.ps1 [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        [switch]$IncludeSystemDbs,
+        [parameter(ValueFromPipeline)]
+        [Object[]]$InputObject,
+        [switch]$EnableException
+    )
+
+    process {
+        if (Test-FunctionInterrupt) {
+            return
+        }
```
Comment (wsmelton) [inline functions/Remove-DbaDbRole.ps1]: Remove

File: functions/Remove-DbaDbRoleMember.ps1 [hunk 4]
```diff
@@ -0,0 +1,156 @@
+#ValidationTags#CodeStyle, Messaging, FlowControl, Pipeline#
```
Comment (wsmelton) [inline functions/Remove-DbaDbRoleMember.ps1]: remove

File: functions/Remove-DbaDbRoleMember.ps1 [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        [string[]]$User,
+        [parameter(ValueFromPipeline)]
+        [Object[]]$InputObject,
+        [switch]$EnableException
+    )
+
+    process {
+        if (Test-FunctionInterrupt) {
+            return
+        }
```
Comment (wsmelton) [inline functions/Remove-DbaDbRoleMember.ps1]: remove

File: tests/Remove-DbaDbRole.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        $dbname1 = "dbatoolsci_$(Get-Random)"
+
+        $null = New-DbaDatabase -SqlInstance $script:instance2 -Name $dbname1 -Owner sa
+        #$null = New-DbaDatabase -SqlInstance $script:instance2 -Name $dbname2 -Owner sa
+
+
+        #$null = $server.Query("CREATE ROLE $role1", 'msdb')
+    }
+    AfterAll {
+        $null = Remove-DbaDatabase -SqlInstance $script:instance1 -Database $dbname1 -confirm:$false
```
Comment (wsmelton) [inline tests/Remove-DbaDbRole.Tests.ps1]: Should be referencing instance2

Also, commented out code should be removed once testing is completed.

**Comment by wsmelton** [comment]
Future reference....PR reviews are much more efficient and go a bit faster if we keep them to one command per PR. Regardless of the dependency this can still be done.

**Comment by dbaduck** [comment]
Totally understand and I will only put one command in each PR in the future.

**Comment by potatoqualitee** [comment]
sweet, looks like all of shawn's concerns have been addressed. thank you all 🏆 👍 💯 ! this will be published to the gallery in 29 days.

**Pull request merged**